tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightTestsBase.cs

```diff
@@ -11,20 +12,37 @@ public class PlaywrightTestsBase : IClassFixture<DashboardServerFixture>, IClass
     public DashboardServerFixture DashboardServerFixture { get; }
     public PlaywrightFixture PlaywrightFixture { get; }
 
+    private readonly ITestOutputHelper _output;
     private IBrowserContext? _context;
 
-    public PlaywrightTestsBase(DashboardServerFixture dashboardServerFixture, PlaywrightFixture playwrightFixture)
+    public PlaywrightTestsBase(DashboardServerFixture dashboardServerFixture, PlaywrightFixture playwrightFixture, ITestOutputHelper output)
     {
         DashboardServerFixture = dashboardServerFixture;
         PlaywrightFixture = playwrightFixture;
+        _output = output;
     }
 
-    public async Task RunTestAsync(Func<IPage, Task> test)
+    public async Task RunTestAsync(Func<IPage, Task<bool>> setup, Func<IPage, Task> test)
     {
         var page = await CreateNewPageAsync();
         try
         {
-            await test(page);
+            var shouldRunTest = true;
+
+            try
+            {
+                await setup(page);
+            }
+            catch (PlaywrightException e)
```

Comment:
Could you please help me understand the fix here. What exception is being thrown here? And how does this try/catch fix the issue?

Reply:
There is a timeout waiting for an element on the dashboard to be ready; I am unable to reproduce it locally. If this occurs during test setup, and not during test execution, we can skip the test.

Reply:
This would skip the test, and we'll be missing the coverage it provides.

Looking at the trace from the issue:
```
  - LocatorAssertions.ToBeVisibleAsync with timeout 5000ms
  - waiting for GetByText("TestResource")
  Stack Trace:
     at Microsoft.Playwright.Core.AssertionsBase.ExpectImplAsync(String expression, FrameExpectOptions expectOptions, Object expected, String message) in /_/src/Playwright/Core/AssertionsBase.cs:line 84
   at Microsoft.Playwright.Core.AssertionsBase.ExpectImplAsync(String expression, ExpectedTextValue[] expectedText, Object expected, String message, FrameExpectOptions options) in /_/src/Playwright/Core/AssertionsBase.cs:line 64
   at Aspire.Dashboard.Tests.Integration.Playwright.PlaywrightFixture.GoToHomeAndWaitForDataGridLoad(IPage page) in /_/tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs:line 28
   at Aspire.Dashboard.Tests.Integration.Playwright.AppBarTests.<AppBar_Change_Theme>b__1_0(IPage page) in /_/tests/Aspire.Dashboard.Tests/Integration/Playwright/AppBarTests.cs:line 25
   at Aspire.Dashboard.Tests.Integration.Playwright.PlaywrightTestsBase.RunTestAsync(Func`2 test) in /_/tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightTestsBase.cs:line 27
   at Aspire.Dashboard.Tests.Integration.Playwright.PlaywrightTestsBase.RunTestAsync(Func`2 test) in /_/tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightTestsBase.cs:line 31
   at Aspire.Dashboard.Tests.Integration.Playwright.AppBarTests.AppBar_Change_Theme() in /_/tests/Aspire.Dashboard.Tests/Integration/Playwright/AppBarTests.cs:line 23
--- End of stack trace from previous location ---
```

The failure is when `GoToHomeAndWaitForDataGridLoad` is waiting for the data grid to show up on the page, and it times out. This could be because the page was loading slowly, or something failed on the page causing the grid to never show up. I would suggest checking for that.

https://github.com/dotnet/aspire/blob/f943b528bac4c88dcca10a3a26a720358dc5b15f/tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs#L25-L31

 Maybe take a screenshot when it times out like this. See https://github.com/dotnet/aspire/blob/f943b528bac4c88dcca10a3a26a720358dc5b15f/tests/Aspire.Workload.Tests/StarterTemplateRunTestsBase.cs#L80-L81

Also, that method is directly trying to open the dashboard url, but it could take time to become available too. We should wait till the url becomes available, see https://github.com/dotnet/aspire/blob/f943b528bac4c88dcca10a3a26a720358dc5b15f/tests/Shared/WorkloadTesting/AspireProject.cs#L298-L304

Reply:
I will re-add a wait function, I had removed it after I received feedback to do so.

Reply:
Actually, just increasing the timeout significantly might be good enough.